builder.go

```diff
@@ -0,0 +1,92 @@
+package cid
+
+import (
+	mh "github.com/multiformats/go-multihash"
+)
+
+type Builder interface {
+	Sum(data []byte) (*Cid, error)
+	GetCodec() uint64
+	WithCodec(uint64) Builder
+}
+
+type V0Builder struct{}
+
+type V1Builder struct {
+	Codec   uint64
+	HashFun uint64
+	HashLen int // HashLen <= 0 means the default length
+}
+
+func PrefixToBuilder(p Prefix) Builder {
+	if p.Version == 0 {
+		return V0Builder{}
+	}
+	mhLen := p.MhLength
+	if p.MhType == mh.ID {
+		mhLen = 0
+	}
+	if mhLen < 0 {
+		mhLen = 0
+	}
+	return V1Builder{
+		Codec:   p.Codec,
+		HashFun: p.MhType,
+		HashLen: mhLen,
+	}
+}
+
+func (p Prefix) GetCodec() uint64 {
+	return p.Codec
+}
+
+func (p Prefix) WithCodec(c uint64) Builder {
```

Comment:
Should comment that this will auto-upgrade to CIDv1 if necessary. Actually, I wonder if this should return a prefix or just return a builder (V1/V0).

Reply:
Yes that is intentional.  I don't see any reason why this should return a builder V1/V0.

Reply:
It was just a thought. I don't really see any benefit one way or the other.